modules/core/include/visp3/core/vpImageConvert.h

```diff
@@ -109,6 +109,9 @@ class VISP_EXPORT vpImageConvert
   static void createDepthHistogram( const vpImage< uint16_t > &src_depth, vpImage< vpRGBa > &dest_rgba );
   static void createDepthHistogram( const vpImage< uint16_t > &src_depth, vpImage< unsigned char > &dest_depth );
 
+  static void createDepthHistogram( const vpImage< float > &src_depth, vpImage< vpRGBa > &dest_depth );
+  static void createDepthHistogram( const vpImage< float > &src_depth, vpImage< unsigned char > &dest_depth );
```

Comment:
Wondering if it is a good idea to add these overloads for `float` type?

Looks like the only difference with `uint16_t` is the handling of `NaN` for incorrect/not available depth.
I don't know if it would be possible to have a `createDepthHistogram` for `float` type but with a very different behaviour?
Maybe renaming the function to something specific to Occipital?

Or split the function behaviour between detecting `NaN`, replacing this value to `0`, converting the `vpImage` to `uint16_t` and calling the `createDepthHistogram` with `uint16_t` type?

Reply:
Thanks @s-trinh for these comments!

In fact, together with @fspindle, we had a similar discussion before implementing these overloads.
Another option was to create a `convert` function from `float` to `uint16_t` that deals with `NaN`. But this `convert` will be just used for creating depth histogram (since depth is returned as float* from the SDK, and we just need to transform this information to a displayed image `vpRGBa/unsigned char`). 

So should we be creating a temporary `uint16_t` image and pass it to the already existing `createDepthHistogram` or create a `createDepthHistogram` for `float` images?

Reply:
For me it is not an issue to have an overloaded function for float image. I prefer this option than introducing a conversion from `float` to `uint16_t` images that is computationnaly costly.

To avoid code duplication, maybe we could create a template function that does the job and that is called in `vpImageConvert::createDepthHistogram()`

I suggest to create the template in `src/image/private/vpImageConvert_impl.h` with something like (not tested)

```
/*!
  Convert the input float depth image to a 8-bits depth image. The input
  depth value is assigned a value proportional to its frequency.
  \param[in] src_depth : Input depth image.
  \param[out] dest_depth : Output grayscale depth image.
  \param[in] check_isNaN : When true test if input depth image values are NaN. 
  If this is the case these pixels are set to 0 in the output image.
 */
template <class Type> 
void
vp_createDepthHistogram( const vpImage< Type > &src_depth, vpImage< unsigned char > &dest_depth, bool check_isNaN )
{
  dest_depth.resize( src_depth.getHeight(), src_depth.getWidth() );
  uint32_t histogram[0x10000];
  memset( histogram, 0, sizeof( histogram ) );

  if (check_isNaN) {
    for ( unsigned int i = 0; i < src_depth.getSize(); ++i ) {
      if(!vpMath::isNaN(src_depth.bitmap[i])) {
        ++histogram[static_cast<uint32_t>(src_depth.bitmap[i])];
      }
    }
  }
  else {
    for ( unsigned int i = 0; i < src_depth.getSize(); ++i ) {
      ++histogram[static_cast<uint32_t>(src_depth.bitmap[i])];
    }
  }

  for ( int i = 2; i < 0x10000; ++i )
    histogram[i] += histogram[i - 1]; // Build a cumulative histogram for the indices in [1,0xFFFF]

  for ( unsigned int i = 0; i < src_depth.getSize(); ++i )
  {
    uint16_t d = src_depth.bitmap[i];
    if ( d )
    {
      unsigned char f =
          static_cast< unsigned char >( histogram[d] * 255 / histogram[0xFFFF] ); // 0-255 based on histogram location
      dest_depth.bitmap[i] = f;
    }
    else
    {
      dest_depth.bitmap[i] = 0;
    }
  }
}
```
where I suppressed the `static` and change the `cast` wrt to current code.

Then in `vpImageConvert.cpp`the code becomes:
```
void
vpImageConvert::createDepthHistogram( const vpImage< float > &src_depth, vpImage< unsigned char > &dest_depth)
{
  vp_createDepthHistogram(src_depth, dest_depth, true);
}

void
vpImageConvert::createDepthHistogram( const vpImage< uint16_t > &src_depth, vpImage< unsigned char > &dest_depth )
{
  vp_createDepthHistogram(src_depth, dest_depth, false);
}
```

Similar thing could be done for `vpRGBa` images.

@s-trinh and @jnader What do you think ?

Reply:
I was just thinking that adding `openmp` support before the `for` iterators could boost the behavior.

Reply:
@fspindle Indeed a template function could be a good solution for avoiding code duplication.

Reply:
I implemented the private template function and ran some performance tests on `createDepthHistogram()` with and without `openmp` support, on both `RealSense D435` and `Structure Core` sensors, and got the following results:

| Hardware | Without OpenMP<br>(Average execution time) | With OpenMP<br>(Average execution time)|
| ------------- | ------------- | ------------- |
| RealSense D435 | 1.64825 ms | 1.34228 ms |
| Structure Core | 2.92123 ms | 0.776386 ms |